Source/DafnyDriver/Commands/TranslateCommand.cs

```diff
@@ -5,37 +5,42 @@
 
 namespace Microsoft.Dafny;
 
-class TranslateCommand : ICommandSpec {
-  public IEnumerable<Option> Options =>
+static class TranslateCommand {
+  public static IEnumerable<Option> Options =>
     new Option[] {
       CommonOptionBag.Output,
       CommonOptionBag.IncludeRuntimeOption,
-    }.Concat(ICommandSpec.TranslationOptions).
-      Concat(ICommandSpec.ConsoleOutputOptions).
-      Concat(ICommandSpec.ResolverOptions);
-
-  private static readonly Argument<string> Target = new("language", @"
-cs - Translate to C#.
-go - Translate to Go.
-js - Translate to JavaScript.
-java - Translate to Java.
-py - Translate to Python.
-cpp - Translate to C++.
-
-Note that the C++ backend has various limitations (see Docs/Compilation/Cpp.md). This includes lack of support for BigIntegers (aka int), most higher order functions, and advanced features like traits or co-inductive types.".TrimStart()
-  );
-
-  public Command Create() {
+    }.Concat(DafnyCommands.TranslationOptions).
+      Concat(DafnyCommands.ConsoleOutputOptions).
+      Concat(DafnyCommands.ResolverOptions);
+
+  public static Command Create() {
     var result = new Command("translate", "Translate Dafny sources to source and build files in a specified language.");
-    result.AddArgument(Target);
-    result.AddArgument(ICommandSpec.FilesArgument);
-    return result;
-  }
 
-  public void PostProcess(DafnyOptions dafnyOptions, Options options, InvocationContext context) {
-    dafnyOptions.Compile = false;
-    var noVerify = dafnyOptions.Get(BoogieOptionBag.NoVerify);
-    dafnyOptions.CompilerName = context.ParseResult.GetValueForArgument(Target);
-    dafnyOptions.SpillTargetCode = noVerify ? 3U : 2U;
+    result.AddCommand(new Command("cs", "C#"));
```

Comment:
Shouldn't the subcommands be dynamically populated based on the set of available backends? Or are you going to leave that for #4591? (That would be just fine if so)

Reply:
There's a catch-22 there because plugins can add new back-ends. I think it would be better if plugins that expose new back-ends are new binaries that reference Dafny as a library. They could potentially then be called using `dafny translate <new-backend-name> by being available on path with the name `dafny-translate-<new-backend-name>`, similar to how `dotnet` does it.

@MikaelMayer

Reply:
Ah that is an unfortunate circular dependency. That sounds like this PR breaks the workflow for compiler plugins though, because something like `dafny translate blort ... --plugin:BlortCompiler.dll` will no longer work?

Reply:
The whole reason why plugins are provided to Dafny rather than the other way round is that it makes it possible to add these plugins to Dafny in VSCode too, including plugins that expose backends, because these backends could also serve as a basis for verifying things specifics to some backends in the future;

I suggest that, for custom back-ends, we simply break the circular dependency by having a special command that emulates the previous behavior, like "custom-backend" that will support an option like "--name:rs" and load a back-end from the set of plugins.

Reply:
> That sounds like this PR breaks the workflow for compiler plugins though, because something like dafny translate blort ... --plugin:BlortCompiler.dll will no longer work?

That's correct, but I don't think someone was doing this.

> The whole reason why plugins are provided to Dafny rather than the other way round is that it makes it possible to add these plugins to Dafny in VSCode too

That you can still do. `dafny server --plugin:AddWhateverYouLikeToTheIDE.dll`

> I suggest that, for custom back-ends, we simply break the circular dependency by having a special command that emulates the previous behavior, like "custom-backend" that will support an option like "--name:rs" and load a back-end from the set of plugins.

I suggest we consider adding that when there's a need, although then I would argue we shouldn't add it.

Reply:
> That's correct, but I can't imagine anyone was doing this.

To clarify: you don't believe anyone was doing this or you just don't know about this use case?

> That you can still do. `dafny server --plugin:AddWhateverYouLikeToTheIDE.dll`

But then they can't choose the back end. I know of at least one team that has two custom back-ends and use VSCode with their custom plugin. I'm pretty sure they will appreciate being able to right-click, and select dafny run, and with their custom arguments can choose which backend it would run on.

> I suggest we consider adding that when there's a need.

This PR is _removing_ what was plugins enabled in the first place, for which @cpitclaudel spent quite some time refactoring the codebase. In general, I would say that any PR that removes a feature without offering workarounds, when the author does not have evidence will not affect customers, should be approved by at least two team members, one of which participated in the creation of the said feature if possible.

Reply:
+1 to what @MikaelMayer is saying, unfortunately. Is there any way we can special-case the `--plugin` option so it can be at least partially processed before resolving the subcommands? I doubt it without moderate hackery but I have to ask anyway. :P

Alternatively, how hard would it be to implement this idea instead? "They could potentially then be called using dafny translate <new-backend-name> by being available on path with the name dafny-translate-<new-backend-name>, similar to how dotnet does it." Then we would at least have a quick workaround to the breaking change.

Reply:
Made a related PR: https://github.com/dafny-lang/dafny/pull/4627

Reply:
> Is there any way we can special-case the --plugin option so it can be at least partially processed before resolving the subcommands?

We can but I don't think we should. Actually I think some of our current back-ends should become separate binaries at some point.

> Alternatively, how hard would it be to implement this idea instead? "They could potentially then be called using dafny translate by being available on path with the name dafny-translate-, similar to how dotnet does it." Then we would at least have a quick workaround to the breaking change.

The 'workaround' is to build a binary instead of a library when building your custom back-end. The users that I know whom are using a custom back-ends already do this. They also use a Dafny plugin but they don't use it for translation to a target language, only to customize resolution of the language server, which this PR still allows.

> But then they can't choose the back end. I know of at least one team that has two custom back-ends and use VSCode with their custom plugin. I'm pretty sure they will appreciate being able to right-click, and select dafny run, and with their custom arguments can choose which backend it would run on.

You can configure tasks in VS code to easily run a custom binary on the current particular file.